flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/CheckedThread.java

```diff
@@ -18,26 +18,42 @@
 
 package org.apache.flink.core.testutils;
 
+import java.util.concurrent.TimeoutException;
+
 /**
  * A thread that additionally catches exceptions and offers a joining method that re-throws the
  * exceptions.
  *
- * <p>Rather than overriding {@link Thread#run()} (or supplying a {@link Runnable}), one needs to
- * extends this class and implement the {@link #go()} method. That method may throw exceptions.
+ * <p>This class needs to supply a {@link RunnableWithException} that may throw exceptions or
+ * override {@link #go()} method.
+ *
+ * <p>you can use it as the same way of using threads like: {@code new Thread(Runnable runnable)} or
+ * {@Code new Thread()} and then override {@link Thread#run()} method. Just change it to {@code new
+ * CheckedThread(RunnableWithException runnableWithException)} or {@Code new CheckedThread()} and
+ * then override {@link CheckedThread#go()} method.
  *
- * <p>Exception from the {@link #go()} method are caught and re-thrown when joining this thread via
- * the {@link #sync()} method.
+ * <p>Exception from the {@link #runnable} or the override {@link #go()} are caught and re-thrown
+ * when joining this thread via the {@link #sync()} method.
  */
-public abstract class CheckedThread extends Thread {
+public class CheckedThread extends Thread {
```

Comment:
we can make this final, remove `go()` and for force `runnable` to be non-null.

Reply:
I think when using `CheckedThread` you should keep the habit of using `Thread`: pass in a runnable or override the go method.
Another reason is that I found that many test classes such as: `org.apache.flink.core.fs.LimitedConnectionsFileSystemTest.BlockingWriterThread` will be ugly if they do not support overriding the go method: I thought about introducing a class `BlockingWriterRunnable` extends `RunnableWithExcpetion`, and then let the field of BlockingWriterThread and the logic of the go method are transferred to BlockingWriterRunnable, and then replaced it with `new CheckedThrad(new BlockingWriterRunnable(xxx) )` where `BlockingWriterThread` is used. This is very inconvenient to use, because the test class will call the `sync` on `CheckedThread`, and also call the `wakeup` method on `BlockingWriterRunnable`,  but in the original implementation, these methods are all on `BlockingWriterThread`

Reply:
The BlockingWriterThread case can be solved quite easily with a factory method and passing the latch via the constructor.

```
    static BlockingWriterThread createBlockingWriterThread(
            LimitedConnectionsFileSystem fs,
            Path path,
            int maxConcurrentOutputStreams,
            int maxConcurrentStreamsTotal) {
        final OneShotLatch waiter = new OneShotLatch();

        return new BlockingThread(
                () -> {
                    try (FSDataOutputStream stream = fs.create(path, WriteMode.OVERWRITE)) {
                        assertTrue(fs.getNumberOfOpenOutputStreams() <= maxConcurrentOutputStreams);
                        assertTrue(fs.getTotalNumberOfOpenStreams() <= maxConcurrentStreamsTotal);

                        final Random rnd = new Random();
                        final byte[] data = new byte[rnd.nextInt(10000) + 1];
                        rnd.nextBytes(data);
                        stream.write(data);

                        waiter.await();

                        // try to write one more thing, which might/should fail with an I/O
                        // exception
                        stream.write(rnd.nextInt());
                    }
                },
                waiter);
```

Reply:
Nice idea, I tried to get rid of the support for overriding the go method, and then found that a more difficult class might not be easy to deal with. It is `org apache. flink. streaming. runtime. io. benchmark. ReceiverThread`,  do you have any suggestions?

Reply:
After reading your opinion, I now think it is appropriate to remove the `go` method, so that our class will become very simple and easy to use.